src/CodeAnalysis/CSharp/MetaAnalyzers/CSharpImmutableObjectMethodAnalyzer.cs

```diff
@@ -0,0 +1,113 @@
+            isEnabledByDefault: true,
+            description: s_localizableDescription,
+            customTags: WellKnownDiagnosticTags.Telemetry);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics
+        {
+            get
+            {
+                return ImmutableArray.Create(DoNotIgnoreReturnValueDiagnosticRule);
+            }
+        }
+
+        private static readonly string s_solutionFullName = @"Microsoft.CodeAnalysis.Solution";
+        private static readonly string s_projectFullName = @"Microsoft.CodeAnalysis.Project";
+        private static readonly string s_documentFullName = @"Microsoft.CodeAnalysis.Document";
+        private static readonly string s_syntaxNodeFullName = @"Microsoft.CodeAnalysis.SyntaxNode";
+
+        // A list of known immutable object names
+        private static ImmutableArray<string> s_immutableObjectNames
+        {
+            get
+            {
+                return ImmutableArray.Create(s_solutionFullName, s_projectFullName, s_documentFullName, s_syntaxNodeFullName);
+            }
+        }
+
+        private static readonly string s_Add = "Add";
+        private static readonly string s_Remove = "Remove";
+        private static readonly string s_Replace = "Replace";
+        private static readonly string s_With = "With";
+
+        private static ImmutableArray<string> s_immutableMethodNames 
+        {
+            get
+            {
+                return ImmutableArray.Create(s_Add, s_Remove, s_Replace, s_With);
+            }
+        }
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.RegisterSyntaxNodeAction(AnalzyeInvocationForIgnoredReturnValue, SyntaxKind.InvocationExpression);
+        }
+
+        public void AnalzyeInvocationForIgnoredReturnValue(SyntaxNodeAnalysisContext context)
+        {
+            var model = context.SemanticModel;
+            var candidateInvocation = (InvocationExpressionSyntax)context.Node;
+
+            //We're looking for invocations that are children of a statement but not children of a return statement.
+            if(!(candidateInvocation.Parent is StatementSyntax) || candidateInvocation.Parent is ReturnStatementSyntax)
+            {
+                return;
+            }
+
+            //If we can't find the method symbol, quit
+            var methodSymbol = model.GetSymbolInfo(candidateInvocation).Symbol as IMethodSymbol;
+            if (methodSymbol == null)
+            {
+                return;
+            }
+
+            //If the method doesn't start with something like "With" or "Replace", quit
+            string methodName = methodSymbol.Name;
+            if (!s_immutableMethodNames.Any(n => methodName.StartsWith(n)))
+            {
+                return;
+            }
+
+            //If we're not in one of the known immutable types, quit
+            var parentName = methodSymbol.ReceiverType.ToString();
```

Comment:
+1 to this suggestion - comparing the symbols directly would be more efficient. Also, it provides the opportunity to not run the analyzer at all incase those symbols are missing. You might want to do something like this in the Initialize method:

```csharp
void Initialize(context)
{
      context.RegisterCompilationStartAction(compilationContext = >
      {
             var solutionSymbol = compilationContext.Compilation.GetTypeFromMetadataName(s_solutionName);
             ...
             if (solutionSymbol != null || ...)
             {
                  compilationContext.RegisterSyntaxNodeAction(sc => AnalyzeInvocation(solutionSymbol, ...), ...);
             }
      }
}
```